r2-core/src/main/java/com/linkedin/r2/message/BaseMessage.java

```diff
@@ -41,12 +41,25 @@
 
   protected BaseMessage(Map<String, String> headers, List<String> cookies)
   {
-    ArgumentUtil.notNull(headers, "headers");
-    ArgumentUtil.notNull(cookies, "cookies");
-    TreeMap<String, String> tmpHeaders = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
-    tmpHeaders.putAll(headers);
-    _headers = Collections.unmodifiableSortedMap(tmpHeaders);
-    _cookies = Collections.unmodifiableList(new ArrayList<>(cookies));
+    this(headers, cookies, true);
+  }
+
+  protected BaseMessage(Map<String, String> headers, List<String> cookies, boolean validateArgs)
+  {
+    if (validateArgs)
+    {
+      ArgumentUtil.notNull(headers, "headers");
+      ArgumentUtil.notNull(cookies, "cookies");
+      TreeMap<String, String> tmpHeaders = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+      tmpHeaders.putAll(headers);
+      _headers = Collections.unmodifiableSortedMap(tmpHeaders);
+      _cookies = Collections.unmodifiableList(new ArrayList<>(cookies));
+    }
+    else
+    {
+      _headers = headers;
```

Comment:
This changed header order in previous code, may break existing code.

Reply:
Sorry I didn't get what you were alluding to. The existing method signature and behavior is unchanged.

Reply:
I meant if `validateArgs` is true, `_headers` is sorted, but it is not if `validateArgs` is false. This may cause issues for code that handles the R2 request if the code has assumption on the headers passed.

Reply:
I don't think we care about map sorting at all! What we care about is case-insensitivity to follow the http header spec. TreeMap just happens to be a sorted map that can be configured as a case -insensitive map.

The entire point of validateArgs is this: We will not validate/copy anything and trust whatever the subclass sent us if this is false. This is supposed to be carefully used by subclasses who know what they are doing (like us creating a url rewritten request copy here).

The default value of validateArgs is true, and in this mode all the existing validation and copying checks run.